neuralpp/experiments/experimental_inference/exact_belief_propagation.py

```diff
@@ -0,0 +1,37 @@
+from typing import List, Iterable
+
+from neuralpp.experiments.experimental_inference.tree import FactorGraph, SpanningTree
+from neuralpp.inference.graphical_model.representation.factor.product_factor import (
+    Factor, ProductFactor
+)
+from neuralpp.inference.graphical_model.variable.variable import Variable
+from neuralpp.util.group import Group
+
+
+class ExactBeliefPropagation:
+
+    def __init__(self, factors: List[Factor], precompute):
+        self.tree = SpanningTree(FactorGraph(factors, precompute))
+        self.sum_points = {}
+
+    def run(self, query: Variable):
+        self.tree.evaluate(query)
+        return self.msg_from_node(query)
+
+    def msg_from_node(self, node):
+        messages = [self.msg_from_node(v) for v in self.tree.neighbors(node)]
+        if isinstance(node, Variable):
+            self.process_variable_sum_point(node)
+            if not messages:
+                return Group.identity
+        else:
+            messages += [node]
+        return ProductFactor.multiply(messages).atomic_factor().sum_out_variables(self.sum_points.get(node, []))
+
+    def process_variable_sum_point(self, variable: Variable):
```

Comment:
Looking at the code here I understand better how you thought about the issue. I am glad you figured this out! However, I think we can make the code significantly simpler. In particular, this function is accessing `tree.graph`, which I think is not ideal because it assumes a certain internal structure for the spanning tree/graph. Ideally, the algorithm should receive _any_ graph and work on that, without using its internal structure or assuming how it was generated. I think this is a bit trickier and it would be good for us to have a meeting before proceeding.